File: packages/zsh-core-tooldef/src/tools/list.ts
// MIRRORED-IN: zshref-rs/src/tools/list.rs

import type { DocCorpus } from "@carlwr/zsh-core"
import type { DocCategory } from "@carlwr/zsh-core/taxonomy"
import { buildToolDef, type SchemaShape, type ToolDef } from "../tool-defs.ts"
import { listProse } from "./prose.ts"
import { type BaseMatch, entries } from "./shared/entries.ts"
import { categoryShape, type Envelope, mkEnvelope } from "./shared/envelope.ts"
import { clampLimit, limitShape } from "./shared/limits.ts"
import { mkOutputSchema } from "./shared/output-schema.ts"

export interface ListInput {
  readonly category?: DocCategory
  readonly limit?: number
}

export type ListMatch = BaseMatch
export type ListResult = Envelope<ListMatch>

/**
 * Enumerate corpus records (optional category filter). Id/display only — no
 * markdown. `limit=0` → metadata only. Pure; no IO.
 */
export function list(corpus: DocCorpus, input: ListInput): ListResult {
  const limit = clampLimit(input.limit)
  const pool = entries(corpus, input.category)
  return mkEnvelope(pool.slice(0, limit), pool.length)
}

const listShape: SchemaShape<"category" | "limit"> = {
  type: "object",
  properties: {
    category: categoryShape,
    limit: limitShape,
  },
  additionalProperties: false,
}

export const listToolDef: ToolDef = buildToolDef<"category" | "limit">({
  name: "zsh_list",
  prose: listProse,
  shape: listShape,
  outputSchema: mkOutputSchema({}),
  execute: (corpus, input): ListResult =>
    list(corpus, input as unknown as ListInput),
})

File: zshref-rs/src/tools/list.rs
//! `zsh_list` — enumerate corpus records (no `mdBody`), optional category filter.
//! `limit=0` → metadata only (`matchesTotal` nonzero, `matches` empty).

use crate::corpus::{Corpus, DocCategory};
use crate::tools::envelope::{mk_entry, mk_envelope};
use crate::tools::schema::{output_schema, MatchShape, Shape};
use crate::tools::{category_input, prose, Field, Tool, ToolName};
use anyhow::Result;
use serde_json::Value;

pub fn tool(corpus: &Corpus) -> Tool {
    Tool::new(
        ToolName::List,
        prose::list(),
        vec![
            Field::optional(
                "category",
                prose::filter_category(corpus.index),
                Shape::Category,
            ),
            Field::optional("limit", prose::limit(), Shape::Limit),
        ],
        output_schema(&MatchShape::default(), corpus),
        run,
    )
}

pub fn run(input: &Value, corpus: &Corpus) -> Result<Value> {
    let category = category_input(input)?;
    let limit = input.get("limit").and_then(Value::as_u64).unwrap_or(0) as usize;

    let pool = entries(corpus, category);
    let total = pool.len();
    let matches: Vec<Value> = pool.into_iter().take(limit).collect();
    Ok(mk_envelope(matches, total))
}

fn entries(corpus: &Corpus, cat_filter: Option<DocCategory>) -> Vec<Value> {
    corpus
        .categories
        .iter()
        .filter(|cat| cat_filter.is_none_or(|f| cat.name == f))
        .flat_map(|cat| {
            cat.records
                .iter()
                .map(move |rec| mk_entry(cat.name, rec.id(), rec.display(), rec.sub_kind(), None))
        })
        .collect()
}
